### PR TITLE
performSaveHandlerUpdate

### LINKED ISSUE #117: SessionConfig -> setPhpSaveHandler() - don't locate registered save handlers if instance of SessionHandlerInterface
If you pass instance of SessionHandlerInterface to setPhpSaveHandler it is locating registered save handlers in phpinfo(). There is no need for this if it is instance of SessionHandlerInterface.
It can be checked for string before calling $this->locateRegisteredSaveHandlers().

### PR DESCRIPTION
**Pull request by dennybrandes**
- check with isString() before calling locateRegisteredSaveHandlers()
- fixed https://github.com/zendframework/zend-session/issues/117

### CODE REVIEW

File: src/Config/SessionConfig.php [hunk 1]
```diff
@@ -447,40 +447,39 @@ private function locateRegisteredSaveHandlers()
...
+                        'Error setting session save handler module "%s": %s',
+                        $phpSaveHandler,
+                        $this->phpErrorMessage
+                    ));
+                }
+
+                return $phpSaveHandler;
+            }
+
+            if ((! class_exists($phpSaveHandler)
```
Comment (michalbundyra) [inline src/Config/SessionConfig.php]: we can remove one parentheses, these are redundant

File: src/Config/SessionConfig.php [hunk 2]
```diff
@@ -447,40 +447,39 @@ private function locateRegisteredSaveHandlers()
...
+                    || ! (in_array(SessionHandlerInterface::class, class_implements($phpSaveHandler)))
```
Comment (michalbundyra) [inline src/Config/SessionConfig.php]: I see this condition was like that before, but I think it can be simplify by using `is_a` method.

**Review by michalbundyra — COMMENTED** [review]
@dennybrandes First I was quite confused by these changes, and then I've changed view in github pr - to splitted and ignore whitespaces and all became clear.

LGTM, just small two comments.

File: src/Config/SessionConfig.php (hunk 1 above)
Comment (dennybrandes) [reply src/Config/SessionConfig.php]: this is fixed.

File: src/Config/SessionConfig.php (hunk 2 above)
Comment (dennybrandes) [reply src/Config/SessionConfig.php]: I tried, but unit tests are not working after change. Need to investigate more. I will do it in another MR.

Comment (michalbundyra) [reply src/Config/SessionConfig.php]: Have you tried `is_a($phpSaveHandler, SessionHandlerInterface::class, true)` ?
I've seen you've tried `instanceof` but this is not gonna work.

Comment (dennybrandes) [reply src/Config/SessionConfig.php]: I tried with is_a but that will break unit tests. Maybe we can fix it in another PR/issue.

**Comment by dennybrandes** [comment]
@webimpress can you check again please? :-)

File: src/Config/SessionConfig.php (hunk 2 above)
Comment (michalbundyra) [reply src/Config/SessionConfig.php]: I've checked and it works for me:
```php
            if (! class_exists($phpSaveHandler)
                || ! is_a($phpSaveHandler, SessionHandlerInterface::class, true)
            ) {
```

File: src/Config/SessionConfig.php [hunk 3]
```diff
@@ -447,40 +447,39 @@ private function locateRegisteredSaveHandlers()
...
+                        $this->phpErrorMessage
+                    ));
+                }
+
+                return $phpSaveHandler;
+            }
+
+            if (! class_exists($phpSaveHandler)
+                    || ! (in_array(SessionHandlerInterface::class, class_implements($phpSaveHandler)))
+
```
Comment (michalbundyra) [inline src/Config/SessionConfig.php]: Remove empty line above, please

**Review by michalbundyra — APPROVED** [review]

File: src/Config/SessionConfig.php [hunk 4]
```diff
@@ -447,40 +447,39 @@ private function locateRegisteredSaveHandlers()
...
+            ) {
                 throw new Exception\InvalidArgumentException(sprintf(
-                    'Error setting session save handler module "%s": %s',
+                    'Invalid save handler specified ("%s"); must be one of [%s]'
+                    . ' or a class implementing %s',
                     $phpSaveHandler,
-                    $this->phpErrorMessage
+                    implode(', ', $knownHandlers),
+                    SessionHandlerInterface::class,
+                    SessionHandlerInterface::class
```
Comment (michalbundyra) [inline src/Config/SessionConfig.php:477]: Above line can be removed, too many params.

File: src/Config/SessionConfig.php (hunk 2 above)
Comment (dennybrandes) [reply src/Config/SessionConfig.php]: You are right, Thanks. Its changed

File: src/Config/SessionConfig.php (hunk 3 above)
Comment (dennybrandes) [reply src/Config/SessionConfig.php]: I removed the empty line

File: src/Config/SessionConfig.php (hunk 4 above)
Comment (dennybrandes) [reply src/Config/SessionConfig.php:477]: I removed it

**Comment by dennybrandes** [comment]
@webimpress i added my change now to changelog, but maybe you need to change the release date there :-)

**Comment by michalbundyra** [comment]
Thanks, @dennybrandes !

**Pull request merged**